jax/random.py

```diff
@@ -876,20 +880,43 @@ def _case4(zagf):
     return grad
 
 def _gamma_grad(sample, a):
-    samples = np.reshape(sample, -1)
-    alphas = np.reshape(a, -1)
+  samples = np.reshape(sample, -1)
+  alphas = np.reshape(a, -1)
+  if xla_bridge.get_backend().platform == 'cpu':
+    grads = lax.map(lambda args: _gamma_grad_one(*args), (samples, alphas))
+  else:
     grads = vmap(_gamma_grad_one)(samples, alphas)
-    return grads.reshape(onp.shape(a))
+  return grads.reshape(onp.shape(a))
 
-@custom_transforms
 def _gamma_impl(key, a):
-    alphas = np.reshape(a, -1)
-    keys = split(key, onp.size(alphas))
+  a_shape = np.shape(a)
+  # split key to match the shape of a
+  key_ndim = np.ndim(key) - 1
+  key = np.reshape(key, (-1, 2))
+  key = vmap(split, in_axes=(0, None))(key, prod(a_shape[key_ndim:]))
+  keys = np.reshape(key, (-1, 2))
+  alphas = np.reshape(a, -1)
+  if xla_bridge.get_backend().platform == 'cpu':
+    samples = lax.map(lambda args: _gamma_one(*args), (keys, alphas))
+  else:
     samples = vmap(_gamma_one)(keys, alphas)
-    return np.reshape(samples, onp.shape(a))
-
-defjvp(_gamma_impl, None,
-       lambda tangent, ans, key, a, **kwargs: tangent * _gamma_grad(ans, a))
+  return np.reshape(samples, a_shape),
+
+def _gamma_batching_rule(batched_args, batch_dims):
+    k, a = batched_args
+    bk, ba = batch_dims
+    size = next(t.shape[i] for t, i in zip(batched_args, batch_dims) if i is not None)
+    k = batching.bdim_at_front(k, bk, size)
+    a = batching.bdim_at_front(a, ba, size)
+    return random_gamma_p.bind(k, a), (0,)
+
+random_gamma_p = core.Primitive('random_gamma')
+random_gamma_p.multiple_results = True
```

Comment:
I think you probably want this to be False (the default), and then you don't need to return tuples elsewhere (e.g. from `_gamma_impl` you can remove the trailing comma on the return line, from `gamma` you can remove the trailing `[0]` on the return line, from the abstract eval rule you can just return the result of `raise_to_shaped` rather than tupling it, etc.).

Reply:
Yeah, at first I didn't use tuple but multiple issues happen (see my topic comment). @jekbradbury correctly pointed out that `xla.lower_fun` enjoys tuple outputs. I'll change it back to see if the issue still happens.

Reply:
Oh, I didn't think of that... though if it's the only problem, I can fix `xla.lower_fun` to allow singleton outputs. I'm curious if there are other issues.

But we can follow up on that later. There's no reason to block merging this now, just for some singleton tuple stuff. Thanks for explaining!

Reply:
I think that is the only issue (see a simple reproducible script in slack channel). I just remove tuple stuff but the issue still happens. :)